src/journal/test-catalog.c

```diff
@@ -202,14 +201,27 @@ static void test_catalog_file_lang(void) {
 
 int main(int argc, char *argv[]) {
         _cleanup_(unlink_tempfilep) char database[] = "/tmp/test-catalog.XXXXXX";
-        _cleanup_free_ char *text = NULL;
+        _cleanup_free_ char *text = NULL, *exedir = NULL, *catalog_dir = NULL;
         int r;
 
         setlocale(LC_ALL, "de_DE.UTF-8");
 
+        log_set_max_level(LOG_DEBUG);
         log_parse_environment();
         log_open();
 
+        /* If test-catalog is located at the build directory, then use catalogs in that.
+         * If it is not, e.g. installed by systemd-tests package, then use installed catalogs. */
+        assert_se(readlink_and_make_absolute("/proc/self/exe", &exedir) >= 0);
+        if (path_startswith(exedir, ABS_BUILD_DIR)) {
```

Comment:
You might want to refactor this code into a shared function such as `bool is_test_running_from_builddir()`, which can then be used in the two places that want to detect that:

https://github.com/systemd/systemd/blob/da0da5eccf5f20874111e0681b111704bf9a1c92/src/shared/tests.c#L38-L41

I'm not super thrilled about using `ABS_BUILD_DIR` to detect this, but given this is pre-existing, it's probably fine to keep detecting it this way. I'm wondering if there's a better way to detect this, such as an environment that gets exported by `ninja` etc., but it's possible this would be called stand-alone on the test environment too, so maybe this is as good as we can do it...

Cheers,
Filipe

Reply:
@filbranden I would very much like to see this. Passing the build directory via an env var when running "ninja test" and not encoding this in the binary itself, would help with making the reproducible. See also my comments in
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=908365#42

One downside of this would be, that you'd have to remember to set that env var when you execute the test-binary manually.

Reply:
The Debian reproducibility effort operates on binary packages. If we drop systemd-tests, there would be not need (from the Debian POV) to get rid of the use of ```ABS_BUILD_DIR```.

IIRC you can declare an autopkgtest as both ```needs_root``` and ```needs_build```. The later would result in the package being built as part of the autopkgtest run. So theoretically this should be possible. Keep in mind though, that a full build requires a lot of ressurces (especially the Debian build which builds both a udeb and non-udeb variant).
Also keep in mind, that whenever a (reverse)dependency of systemd changes, the Debian CI (and I think the Ubuntu CI as well) will trigger a full autopkgtest run [1].
If that would include a full rebuild each time, this would suddenly get a lot more expensive.
Installed tests are useful if you want to execute them often.

[1]  https://ci.debian.net/packages/s/systemd/testing/amd64/

Reply:
being able to run the tests as root is one aspect, being able to run them often (as explained above) is another

Reply:
Maybe https://wiki.gnome.org/Initiatives/GnomeGoals/InstalledTests does a better job at explainging it

Reply:
I'm not sure I understand what was explained above. Who needs to run the unit tests often and why? Nobody had needed that before f853c6efb509589c8e3a82caa3af2 was merged. Probably I missed what happened afterwards.

Reply:
I'm sorry. I didn't see the link you posted when I was writing the previous comment. I'll try to find the answer there. Or, more likely, I'll find what is written there unconvincing :-)

Reply:
Let me get this right. We complicated the tests because for some reason it wasn't possible to just run:
```sh
meson
ninja ... test
sudo ninja ... test
```
on Ubuntu CI, which, in turn, apparently broke the reproducibility of the Debian package. And all we have now is the ability to run the tests often according to the proposal that hasn't been approved yet and, judging by the absence of the files with the `[Test]` section in https://salsa.debian.org/systemd-team/systemd, this ability isn't quite useful at the moment.

Reply:
Those .test files are only necessary if you want to run the tests using GLibs test runner called gtester. We don't use those. The link to the GNOME proposal was just to show the motivation for installed tests

Reply:
anyway, this is really getting OT for this PR

Reply:
I wouldn't say that discussing the underlying issue is off-topic, but, if everybody is OK to complicate things by adding the unnecessary complexity, which is, in my opinion, very distro-specific, then, well, it's fine. It should be added somewhere.

Reply:
There is nothing distro-specific about using installed tests. But I feel like we are going in circles, so I'll stop at this point.

Reply:
I agree that there's nothing distro-specific in installed tests per se. However, to be really useful outside of the environments where there are special scripts like https://salsa.debian.org/systemd-team/systemd/blob/master/debian/tests/root-unittests, the tests should come with some kind of a runner which would run them. Without it they aren't that useful, which makes me think that they aren't widely used. Though, I may be wrong of course. Anyway, I'll stop stop here too. Thank you for the links. I've learnt something new today.